### ISSUE TITLE
PouchDB with Nuxt 3: Uncaught ReferenceError: global is not defined

### ISSUE THREAD

**Issue by Extarys** [pouchdb/pouchdb#8516]
### Issue
Getting `Uncaught ReferenceError: global is not defined` with the line: `var Mutation = global.MutationObserver || global.WebKitMutationObserver;` in file `node_modules/immediate/lib/mutation.js` when using Nuxt 3.

Not sure I provided enough details... I also created a [SO question](https://stackoverflow.com/questions/72483284/how-to-use-rxdb-with-pouchdb-adapter-in-a-nuxt-3-project) for RxDB when used with PouchDB.

My main goal is to use RxDB, but I need a way to make PouchDB works first.

### Info
- Environment: Browser, Nuxt 3 RC 3
- Platform: FF
- Adapter: idb
- Server: CouchDB

### Reproduce

1. `npx nuxi init`
2. `npm install pouchdb`
3. `plugin/pouchdb.ts`:
```
import PouchDB from "pouchdb"

export default defineNuxtPlugin(async nuxtApp => {

    var db = new PouchDB('dbname');

    db.put({
        _id: 'dave@gmail.com',
        name: 'David',
        age: 69
    });

    db.changes().on('change', function () {
        console.log('Ch-Ch-Changes');
    });

    return {
        provide: {
            db: () => db
        }
    }
});
```

I tried adding the following to `nuxt.config.ts` 
```
build: {
		transpile: [
			'pouchdb-browser',
			'pouchdb-utils',

		],
	},
```
I'm far from an expert with transpiling and webpack and such. I saw some polyfill with webpack for the `global` error [here](https://github.com/pouchdb/pouchdb/issues/7984#issuecomment-1003587487).

I tried adding it with Nuxt 3 but either Nuxt didn't add it or this isn't the right fix for my issue...

```
	app: {
		head: {
			script: [
				{
					children: `window.process = window.process || { env: { NODE_DEBUG: undefined, DEBUG: undefined } }; window.global = window;`
				}
			]
		}
	},
```

Nuxt config API [here](https://v3.nuxtjs.org/api/configuration/nuxt.config)

**Comment by stjoe4** [comment]
They can't fix this issue.

**Comment by Extarys** [comment]
Oh ok, thanks for your reply. Quite funny, libraries designed for client-side database sync stuff not working with recent tech, PWA and the like :/

I hope someone eventually finds a solution, in the meantime I'll write everything server side with CouchDB Nano and refactor everything later. Or maybe plain pgsql since it's server-side no matter what, idk.

**Comment by colearendt** [comment]
FWIW I used a polyfill to resolve this particular `global` issue here: https://github.com/colearendt/js-issue-20220729

However, I then ran into #8535, which I don't have a solution for 🙈 I'm probably just going to use axios to make the requests I need since I'm leaning on CouchDB anyways.

**Comment by Extarys** [comment]
Ahh RIP. Similar dance on my side, one error fixed and two appears.

I wish it was easier to use CouchDB. I want to use it so badly with RxJS/RxDB, but Postgres/Prisma right now looked like my only option. I'll figure out some logic to download the needed data for offline use. I'll keep a close eye on CouchDB/Cloudant development, though.

But your idea of using axios is intriguing. I need to use server side API anyway, axios could be used there to fetch data and verify ACL/PASETO anyway. Since I'm not far in the project it could be done. idk

**Comment by sibevin** [comment]
Here is a workaround:

1. Use `pouchdb` package (Not `pouchdb-browser`)
2. Add polyfill script by modifying `App.vue` as following:
```html
<template>
  <div>
    <Head>
      <Script
        children="if (window.global === undefined) { window.global = window; }"
      >
      </Script>
    </Head>
    <NuxtPage />
  </div>
</template>
```
versions:
- nuxt: 3.0.0-rc.8
- pouchdb: 7.3.0
- pouchdb-find: 7.3.0

**Comment by sibevin** [comment]
Here is a workaround:

- nuxt: 3.0.0-rc.8
- pouchdb: 7.3.0
- pouchdb-find: 7.3.0

1. Use `pouchdb` package. `pouchdb-browser` can't work because Nuxt / vite SSR would still load packages on the server side, even we only use them on client pages.
2. Add the polyfill script by modifying App.vue as following:
```html
<template>
  <div>
    <Head>
      <Script
        children="if (window.global === undefined) { window.global = window; }"
      >
      </Script>
    </Head>
    <NuxtPage />
  </div>
</template>
```
3. Initialize DB in a "client" plugin (If not, duplicated DBs are initialized on both server and client sides, and it may not what we expect):

_plugins/initAppDb.client.ts_
```ts
import PouchDB from "pouchdb";
import PouchDBFind from "pouchdb-find";
PouchDB.plugin(PouchDBFind);

export default defineNuxtPlugin(async () => {
  const appDb = new PouchDB("myAppDb");
  await appDb.createIndex({
    index: {
      fields: ["type", "title"],
      ddoc: "index-article-title",
    },
  });
  return {
    provide: { appDb },
  };
});
```
4. Use the plugin in pages

```html
<script setup lang="ts">
  const { $appDb } = useNuxtApp();
  //...
</script>
```

**Comment by lokeshparmar998** [comment]
I am trying to use pouchdb with nuxt-3 and faced the same error "global is not defined". I fixed it by changing the builder in nuxt.config.ts from vite (default builder) to webpack 5.

`export default defineNuxtConfig({ builder: 'webpack' })`

Do not forget to install `@nuxt/webpack-builder` as a dev dependency.

And import it like this for typescript: `import * as PouchDB from '@/node_modules/pouchdb/dist/pouchdb'`

**Comment by VladimirCores** [comment]
Any chance we will have a normally working library that can be imported as any other ES-module libraries as `import PouchDB from 'pouchdb';`?

**Comment by lokeshparmar998** [comment]
@VladimirCores checkout my Nuxt 3, PouchDb [template repository ].(https://github.com/lokeshparmar998/Nuxt-3-Capacitor-4-Pouchdb-Template)

**Comment by VladimirCores** [comment]
The thing is that I need setup for ES modules which is used in Vite.

**Comment by catpea** [comment]
@VladimirCores perhaps this will work (it requires **NO** special configuration in vite.config.js): https://www.npmjs.com/package/pouchdb-browser

I ran a quick test under vite+svelte and it seems to work, though I only had a minute to type this up, sorry.

The following is a random bit of code from the script section of a svelte app, showing initialization, plugin use, and some debugging data.


```JavaScript
import { onMount } from 'svelte';
import PouchDB from 'pouchdb-browser';
import pouchdbFind from 'pouchdb-find';

PouchDB.plugin(pouchdbFind);
const db = new PouchDB('kittens');
 
db.changes().on('change', function(x) {
  console.log('Ch-Ch-Changes',x);
});

onMount(async () => {

  try {
    const result = await db.bulkDocs([
      {_id: 'a', type: 'node', name: 'Application' },
      {_id: 'z', type: 'node', name: 'Utils' },
      {_id: 'b', type: 'node', name: 'Library' },
      {_id: 'c', type: 'node', name: 'Program' },
      {_id: 'x', type: 'node', name: 'Get Date' },
      {_id: 'y', type: 'node', name: 'Load Thing' },
      {_id: 'e1', type:'edge',  source: 'a', target: 'b' },
      {_id: 'e2', type:'edge',  source: 'a', target: 'z' },
      {_id: 'e3', type:'edge',  source: 'b', target: 'c' },
      {_id: 'e4', type:'edge',  source: 'c', target: 'x' },
      {_id: 'e5', type:'edge',  source: 'c', target: 'y' },
    ]);
  } catch (err) {
    console.log(err);
  }

  await db.viewCleanup()
  const indexes = await db.getIndexes();
  console.log(indexes);

const nodes = (await db.query(function (doc, emit) { emit(doc.type); }, {key: 'node', include_docs:true }))
.rows.map(row=>row.doc)
.map(doc=>({data:{id:doc._id, ...doc}}));

const edges = (await db.query(function (doc, emit) { emit(doc.type); }, {key: 'edge', include_docs:true }))
.rows.map(row=>row.doc)
.map(doc=>({data:{id:doc._id, ...doc}}));

const data ={
nodes,
edges,
}

console.log(data);
});
```


Snippet from package.json, I am using:

```
    "pouchdb-browser": "^7.3.0",
    "pouchdb-find": "^7.3.0",
```

**Comment by Sjoerd82** [comment]
I ported my project back from Nuxt to a client side SPA (PouchDB > SSR in my project). But Lo And Behold: I still get this same error message. Without doing any SSR!

Mind blows.

What is going on here???

**Comment by adueck** [comment]
I had this **same issue using Vite** to build a SPA. 

I was able to solve it using a *heavily-shimmed* `vite.config.ts` from [this discussion](https://github.com/vitejs/vite/issues/9703)

It seems that the only problem is PouchDb's use of `immediate` which uses `global`. It would be nice if there was another alternative so that we could use PouchDb without friction in modern build pipelines.

**Comment by Wildanzr** [comment]
> I had this **same issue using Vite** to build a SPA with PouchDB. (Using either `pouchdb` or `pouchdb-browser` v8).
> 
> For me, @Sjoerd82 `define: { global: "window" }` was not enough. I was able to solve it my using [this complex `vite.config.ts`](https://github.com/vitejs/vite/issues/9703#issuecomment-1418261670) to shim for `global` and cjs stuff.
> 
> It seems that **the problem is PouchDB's use of `immediate` which uses `global`**. It would be nice if there was another alternative so that we could use PouchDB without friction in modern build pipelines.

Finally it works in vite react , thank you

**Comment by marcodiri** [comment]
I got it working by adding `define: { global: "window" }` to `vite.config.ts` and installing events dependency (`npm i events`) as suggested in https://github.com/pouchdb/pouchdb/issues/8607#issuecomment-1446127547.

**Comment by connorf25** [comment]
> I got it working by adding `define: { global: "window" }` to vite.config.ts and installing events dependency (`npm i events`) as suggested in [#8607 (comment)](https://github.com/pouchdb/pouchdb/issues/8607#issuecomment-1446127547). Edit: I'm using Ionic+Vite and adding `define: { global: "window" }` causes errors on Ionic components when launching the `vite build` command. To fix this I removed it and added `<script>window.global = window</script>` to index.html instead.

This is how I solved it and it also seems to be the best workaround for Vite, thank you!

**Comment by alxndrsn** [comment]
This may be caused by https://github.com/calvinmetcalf/immediate/issues/39.

**Comment by eshanit** [comment]
I you are using nuxt3 and you are getting the error from pouchdb , you can solve this by:
1. Install events
2. in nuxt.config.ts add this -> vite: {
    define: { global: 'window' },
  },

thats it , you are good to go

**Comment by garethbowen** [comment]
Fixed by #8849
